src/file.cpp

```diff
@@ -201,42 +201,60 @@ namespace {
 		}
 
 		int ret = 0;
+		int num_waits = num_bufs;
 		for (int i = 0; i < num_bufs; ++i)
 		{
 			DWORD num_read;
-			if (ReadFile(fd, bufs[i].iov_base, DWORD(bufs[i].iov_len), &num_read, &ol[i]) == FALSE
-				&& GetLastError() != ERROR_IO_PENDING
+			if (ReadFile(fd, bufs[i].iov_base, DWORD(bufs[i].iov_len), &num_read, &ol[i]) == FALSE)
+			{
+				DWORD last_error = GetLastError();
+				if (last_error == ERROR_HANDLE_EOF)
+				{
+					num_waits = i;
+					break;
+				}
+				else if (last_error != ERROR_IO_PENDING
 #ifdef ERROR_CANT_WAIT
-				&& GetLastError() != ERROR_CANT_WAIT
+					&& last_error != ERROR_CANT_WAIT
 #endif
-				)
-			{
-				ret = -1;
-				goto done;
+					)
+				{
+					ret = -1;
+					goto done;
+				}
 			}
 		}
 
-		if (wait_for_multiple_objects(int(h.size()), h.data()) == WAIT_FAILED)
+		if (num_waits == 0)
+		{
+			goto done;
+		}
+
+		if (wait_for_multiple_objects(num_waits, h.data()) == WAIT_FAILED)
 		{
 			ret = -1;
 			goto done;
 		}
 
-		for (auto& o : ol)
+		for (int i = 0; i < num_waits; ++i)
```

Comment:
what would you think about making this a range-for loop like this?

`for (auto& o : span<OVERLAPPED>(ol).first(num_waits))`

Reply:
ok